libs/d3d12/main.c

```diff
@@ -0,0 +1,377 @@
+/*
+ * Copyright 2018 Józef Kucia for CodeWeavers
+ * Copyright 2020 Joshua Ashton for Valve Software
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+ *
+ */
+
+#define VK_NO_PROTOTYPES
+#include <vkd3d_win32.h>
+#include <stdio.h>
+
+#define ERR(fmt, ...) printf("vkd3d: " fmt, ##__VA_ARGS__)
+#define WARN(fmt, ...) printf("vkd3d: " fmt, ##__VA_ARGS__)
+#define TRACE(fmt, ...) printf("vkd3d: " fmt, ##__VA_ARGS__)
+#define FIXME(fmt, ...) printf("vkd3d: " fmt, ##__VA_ARGS__)
+
+static PFN_vkGetInstanceProcAddr load_vulkan(void)
+{
+    static HMODULE vulkan_module = NULL;
+
+    if (!vulkan_module)
+        vulkan_module = LoadLibraryA("vulkan-1.dll");
+
+    if (vulkan_module)
+        return (void *)GetProcAddress(vulkan_module, "vkGetInstanceProcAddr");
+
+    return NULL;
+}
+
+static const char* debugstr_guid(REFIID iid)
+{
+    return "SOMEGUID";
+}
+
+HRESULT WINAPI D3D12GetDebugInterface(REFIID iid, void **debug)
+{
+    TRACE("iid %s, debug %p.\n", debugstr_guid(iid), debug);
+
+    WARN("Returning DXGI_ERROR_SDK_COMPONENT_MISSING.\n");
+    return DXGI_ERROR_SDK_COMPONENT_MISSING;
+}
+
+HRESULT WINAPI D3D12EnableExperimentalFeatures(UINT feature_count,
+        const IID *iids, void *configurations, UINT *configurations_sizes)
+{
+    FIXME("feature_count %u, iids %p, configurations %p, configurations_sizes %p stub!\n",
+            feature_count, iids, configurations, configurations_sizes);
+
+    return E_NOINTERFACE;
+}
+
+static HRESULT d3d12_signal_event(HANDLE event)
+{
+    return SetEvent(event) ? S_OK : E_FAIL;
+}
+
+struct d3d12_thread_data
+{
+    PFN_vkd3d_thread main_pfn;
+    void *data;
+};
+
+static DWORD WINAPI d3d12_thread_main(void *data)
+{
+    struct d3d12_thread_data *thread_data = data;
+
+    thread_data->main_pfn(thread_data->data);
+    free(thread_data);
```

Comment:
Would probably be cleaner to do this on `join`.

Reply:
We cannot free it on join, we don't have that information there, only a thread handle.